[MLIR] Improve translation of DISubrange.

abidh opened the pull request:
The DISubrange can take integer, dwarf expressions or variables. The current translation only handled integers. This PR adds handling of dwarf expressions and variables.

abidh committed:
[MLIR] Improve translation of DISubrange.

The DISubrange can take integer, dwarf expressions or variables. The
current translation only handled integers. This PR adds handling of
dwarf expressions and variables.

Dinistro commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:
```diff
@@ -625,10 +625,10 @@ def LLVM_DINamespaceAttr : LLVM_Attr<"DINamespace", "di_namespace",
-    OptionalParameter<"IntegerAttr">:$lowerBound,
-    OptionalParameter<"IntegerAttr">:$upperBound,
-    OptionalParameter<"IntegerAttr">:$stride
+    OptionalParameter<"::mlir::Attribute">:$count,
```
I do not recall this, but I guess attribute parameters cannot be constraint through table gen, or can they? Ideally, we would explicitly list which types are legal here.

abidh replied:
I could not find a simple way to add that constraint so I am open to suggestions here. My initial approach was to add a `DISubrangeValueAttr` type which has fields for all the types that we expect (e.g. `IntegerAttr`, `DIExpressionAttr`, etc). Then use that in the `DISubrangeAttr` in place of `IntegerAttr`. But it seems a bit overkill to me. Also made the usage a bit more cumbersome.

gysit replied:
The LLVM BoundType is a pointer union. That could be an option that maybe less cumbersome than a separate attribute. Worst case we can also go for a plain attribute.

Dinistro commented on mlir/lib/Target/LLVMIR/DebugTranslation.cpp:
```diff
@@ -314,11 +314,24 @@ llvm::DINamespace *DebugTranslation::translateImpl(DINamespaceAttr attr) {
+                   llvm::dyn_cast<LLVM::DIGlobalVariableAttr>(attr)) {
+      return translate(globalVar);
+    } else {
+      llvm_unreachable("Unexpected Attribute value");
```
This is not verified on the input IR, so it's technically legal to break this assumption. Therefore, you should not use `llvm_unreachable` here.

abidh replied:
I changed it to return nullptr.

Dinistro commented on mlir/lib/Target/LLVMIR/DebugTranslation.cpp:
```diff
@@ -314,11 +314,24 @@ llvm::DINamespace *DebugTranslation::translateImpl(DINamespaceAttr attr) {
-    return llvm::ConstantAsMetadata::get(llvm::ConstantInt::getSigned(
-        llvm::Type::getInt64Ty(llvmCtx), attr.getInt()));
+
+    if (auto intAttr = llvm::dyn_cast<IntegerAttr>(attr)) {
```
Nit: Use a TypeSwitch here. 
Side note: When you would keep this, you should turn all the `else if`s into normal `if`s, given that all of them contain a return.

abidh replied:
Done.

Dinistro commented on mlir/test/Target/LLVMIR/llvmir-debug.mlir:
```diff
@@ -519,3 +519,41 @@ llvm.func @fn_with_composite() {
+
+// Test that Subrange works with expression and variables.
+
+#di_basic_type = #llvm.di_basic_type<tag = DW_TAG_base_type, name = "int">
+#di_file = #llvm.di_file<"debug-info.ll" in "/">
+#di_compile_unit = #llvm.di_compile_unit<id = distinct[1]<>, sourceLanguage = DW_LANG_Fortran95, file = #di_file, isOptimized = false, emissionKind = Full>
+#di_composite_type = #llvm.di_composite_type<tag = DW_TAG_array_type, name = "expr_elements", baseType = #di_basic_type, flags = Vector, elements = #llvm.di_subrange<count = #llvm.di_expression<[DW_OP_push_object_address, DW_OP_plus_uconst(16), DW_OP_deref]>>>
+#di_subroutine_type = #llvm.di_subroutine_type<types = #di_basic_type, #di_composite_type>
+#di_subprogram = #llvm.di_subprogram<id = distinct[2]<>, compileUnit = #di_compile_unit, scope = #di_file, name = "subranges", file = #di_file, subprogramFlags = Definition, type = #di_subroutine_type>
+#di_local_variable = #llvm.di_local_variable<scope = #di_subprogram, name = "size">
+#gv1 = #llvm.di_global_variable<scope = #di_compile_unit, name = "gv", file = #di_file, line = 3, type = #di_basic_type>
+#gve1 = #llvm.di_global_variable_expression<var = #gv1, expr = <>>
+#di_composite_type1 = #llvm.di_composite_type<tag = DW_TAG_array_type, name = "var_elements", baseType = #di_basic_type, flags = Vector, elements = #llvm.di_subrange<count = #di_local_variable, stride = #gv1>>
+#di_local_variable1 = #llvm.di_local_variable<scope = #di_subprogram, name = "size", type = #di_composite_type1>
+#loc1 = loc("test.f90": 1:1)
+#loc2 = loc(fused<#di_subprogram>[#loc1])
```
Are all of these attribute parameters required?
Also, can you try to split the attributes over different lines (as for the first test in this file), to ensure this is somewhat in the 80 lines character limit (which is not properly enforced in this file).

abidh replied:
Testcase is bigger because we need to test with the local and global variables in the type apart from expression. I have modified the test to respect 80 character limit.

Dinistro reviewed (commented):
Thanks for the improvement, I dropped a few nit comments.

abidh committed:
Handle review comments.

1. Handle formatting nits.
2. Use TypeSwitch instead of multiple if/else
3. Respect 80 column limit in tests.

abidh commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -217,21 +217,34 @@ DISubprogramAttr DebugImporter::translateImpl(llvm::DISubprogram *node) {
+    if (data.isNull())
+      return nullptr;
+
+    if (auto *constInt = llvm::dyn_cast<llvm::ConstantInt *>(data)) {
```
Thanks for the review. Please see the comments below why I could not use `TypeSwitch` here..

abidh commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -217,21 +217,34 @@ DISubprogramAttr DebugImporter::translateImpl(llvm::DISubprogram *node) {
+      else if (auto *global = llvm::dyn_cast<llvm::DIGlobalVariable>(var))
+        return translate(global);
+      llvm_unreachable("Unknown variable type");
+    }
```
The `data` is `PointerUnion` having `DIVariable` as one of its members. Direct casting to `DILocalVariable` or `DIGlobalVariable` does not work. For this reason, I could not use `TypeSwitch` here as the `DIVariable` case returns 2 different types which causes build to fail. I have changed the if/else to if as all had returns.

abidh commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -217,21 +217,34 @@ DISubprogramAttr DebugImporter::translateImpl(llvm::DISubprogram *node) {
-  IntegerAttr count = getIntegerAttrOrNull(node->getCount());
-  IntegerAttr upperBound = getIntegerAttrOrNull(node->getUpperBound());
+  auto count = getAttrOrNull(node->getCount());
+  auto upperBound = getAttrOrNull(node->getUpperBound());
```
Done.

Dinistro reviewed (commented):
Thanks for addressing the comments. Dropped a few additional ones, but I assume that this will be ready for merge soonish.

gysit commented:
> @gysit do you know if there is any way to achieve this?

I don't have a great idea either. I think going with an attribute is probably best we can do for now.

abidh committed:
Handle review comments.

abidh commented on mlir/lib/Target/LLVMIR/DebugTranslation.cpp:
```diff
@@ -314,11 +314,29 @@ llvm::DINamespace *DebugTranslation::translateImpl(DINamespaceAttr attr) {
+
+    llvm::Metadata *metadata =
+        llvm::TypeSwitch<Attribute, llvm::Metadata *>(attr)
+            .Case<IntegerAttr>([&](IntegerAttr intAttr) {
```
Thanks for the tip. I have fixed it.

abidh commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -217,21 +217,33 @@ DISubprogramAttr DebugImporter::translateImpl(llvm::DISubprogram *node) {
-  auto getIntegerAttrOrNull = [&](llvm::DISubrange::BoundType data) {
-    if (auto *constInt = llvm::dyn_cast_or_null<llvm::ConstantInt *>(data))
+  auto getAttrOrNull =
+      [&](llvm::DISubrange::BoundType data) -> mlir::Attribute {
```
Thanks for catching this. Fixed now.

abidh commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -217,21 +217,33 @@ DISubprogramAttr DebugImporter::translateImpl(llvm::DISubprogram *node) {
+      [&](llvm::DISubrange::BoundType data) -> mlir::Attribute {
+    if (data.isNull())
+      return nullptr;
+    if (auto *constInt = llvm::dyn_cast<llvm::ConstantInt *>(data))
```
Done.

abidh commented on mlir/test/Target/LLVMIR/llvmir-debug.mlir:
```diff
@@ -519,3 +519,50 @@ llvm.func @fn_with_composite() {
+#loc1 = loc("test.f90": 1:1)
+#loc2 = loc(fused<#sp>[#loc1])
+
+module attributes {} {
```
Done.

gysit reviewed (commented):
Looks good to me.

Can you try if a PointerUnion works for the BoundType? If it is also cumbersome to use then we can go with the current solution.

abidh committed:
Handle review comments.

abidh commented on mlir/test/Target/LLVMIR/llvmir-debug.mlir:
```diff
@@ -519,3 +519,48 @@ llvm.func @fn_with_composite() {
+
+// CHECK: !DICompositeType(tag: DW_TAG_array_type, name: "var_elements"{{.*}}elements: ![[ELEMENTS2:[0-9]+]])
+// CHECK: ![[ELEMENTS2]] = !{![[ELEMENT2:[0-9]+]]}
+// CHECK: ![[ELEMENT2]] = !DISubrange(count: ![[SR2:[0-9]+]], stride: ![[GV:[0-9]+]])
```
Done.

abidh commented on mlir/lib/Target/LLVMIR/DebugImporter.cpp:
```diff
@@ -217,21 +217,32 @@ DISubprogramAttr DebugImporter::translateImpl(llvm::DISubprogram *node) {
-  IntegerAttr count = getIntegerAttrOrNull(node->getCount());
-  IntegerAttr upperBound = getIntegerAttrOrNull(node->getUpperBound());
+  mlir::Attribute count = getAttrOrNull(node->getCount());
+  mlir::Attribute upperBound = getAttrOrNull(node->getUpperBound());
```
Done.

gysit reviewed (approved):
LGTM

I am not sure if the suggestion of using a PointerUnion got lost? I think that may actually be a good solution to improve type safety of the bounds type info and it is also in line with what LLVM does. Or did you try and it did also get to cumbersome to use?

abidh commented:
> Looks good to me.
> 
> Can you try if a PointerUnion works for the BoundType? If it is also cumbersome to use then we can go with the current solution.



> LGTM
> 
> I am not sure if the suggestion of using a PointerUnion got lost? I think that may actually be a good solution to improve type safety of the bounds type info and it is also in line with what LLVM does. Or did you try and it did also get to cumbersome to use?

I spent time on that suggestion today but I have not been able to make it work till now. Build error suggest that `PointerUnion` does not obey constraints that rest of the code is expecting. I guess some customized parsing/printing code needs to be written before it could be used. I also could not find an example of using an arbitrary type as parameter so it has mostly been trial and error.

gysit commented:
> I spent time on that suggestion today but I have not been able to make it work till now. Build error suggest that PointerUnion does not obey constraints that rest of the code is expecting. I guess some customized parsing/printing code needs to be written before it could be used. I also could not find an example of using an arbitrary type as parameter so it has mostly been trial and error.

Thanks for tying this!

You are right this probably needs a custom parser/printer a bit similar to `LLVM_DIParameter`. Unfortunately the current infrastructure does not provide a nice way to implement such sum types. Let's land as is then.
